From d6e833810909a4278f59c19dbc748424ca5e153f Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Zbigniew=20J=C4=99drzejewski-Szmek?= <zbyszek@in.waw.pl>
Date: Tue, 13 Jan 2026 14:14:22 +0100
Subject: [PATCH 3/6] rpm: add fileattr multifile generator

This allows dlopen notes to be turned into appropriate dependencies
automatically. The dlopen_notes.attr file needs to be installed into
%{_fileattrsdir}.

By default, dependencies are generated for all files that have package
notes. I think this is a reasonable default because it makes the whole
feature easier to discover. In more realistic cases, esp. with
multiple subpackages, it's likely that the packager may need to
configure the distribution of dependencies between subpackages.

One shortcoming of the scheme is that everything is per file, so it's
not possible to say that dependencies generated from a feature should
be assigned to a different subpackage. This is how the feature is
designed in rpm.

The opt-out mechanism is a bit clunky. The first option I considered
was to tell the user to undefine
%__dlopen_notes_requires/recommends/suggests, but that requires three
lines of boilerplate. And might not be forwards-compatible if we add
new features in the future. The second option would be to tell the
user to define __dlopen_notes_requires/recommends/suggests_opts to
%nil. But that has similar problems. I think it's nice to have an
obvious oneliner to handle this. Unfortunately, when I tried to use
  %__dlopen_notes_requires %{?_dlopen_notes_generator:%{_dlopen_notes_generator} ...}
  %__dlopen_notes_recommends %{?_dlopen_notes_generator:%{_dlopen_notes_generator} ...}
  %__dlopen_notes_suggests %{?_dlopen_notes_generator:%{_dlopen_notes_generator} ...}
in the .attr file, when the package has %undefine _dlopen_notes_generator,
we still end up with the macro being expanded. Maybe I misunderstood
the macro expansion logic. The approach with 'true' is clunky, but
it works fine.

Thanks to Neal Gompa for the suggestion to use this protocol.

The new interface is new, independent of the existing options
--feature, --rpm-recommends, --rpm-requires that were previously added
to support rpms. Unfortunately, with the fileattr protocol, the
old way to specify information is not useful. Instead of trying
to shoehorn the new metadata into existing options, I think it's
easier to add a new set with clear semantics.
---
 README.md             | 13 ++++++
 dlopen-notes.py       | 96 +++++++++++++++++++++++++++++++++++++++++--
 rpm/dlopen_notes.attr | 44 ++++++++++++++++++++
 3 files changed, 150 insertions(+), 3 deletions(-)
 create mode 100644 rpm/dlopen_notes.attr

diff --git a/README.md b/README.md
index 8b27e19de3..f7fa518630 100644
--- a/README.md
+++ b/README.md
@@ -97,6 +97,19 @@ $ dlopen-notes /usr/lib64/systemd/libsystemd-shared-257.so
 ...
 ```
 
+### Using the rpm fileattr generator
+
+The tool that processes package notes can be hooked into the rpm build process
+to automatically generate virtual `Requires`, `Recommends`, and `Suggests` dependencies.
+
+The rpm file attribute mechanism is described in
+[rpm-dependency-generators.7](https://rpm-software-management.github.io/rpm/man/rpm-dependency-generators.7).
+
+This tool implements the 'multifile' protocol:
+it reads the list of files on stdin and outputs a list of virtual dependencies.
+
+See the `rpm/dlopen_notes.attr` file for invocation details and options.
+
 ## Requirements
 * binutils (>= 2.39)
 * mold (>= 1.3.0)
diff --git a/dlopen-notes.py b/dlopen-notes.py
index f103611c00..4fa30d8e89 100755
--- a/dlopen-notes.py
+++ b/dlopen-notes.py
@@ -7,6 +7,7 @@ Read .note.dlopen notes from ELF files and report the contents.
 
 import argparse
 import enum
+import fnmatch
 import functools
 import json
 import sys
@@ -84,6 +85,16 @@ class Priority(enum.Enum):
     def __lt__(self, other):
         return self.value < other.value
 
+    def rpm_name(self):
+        if self == self.__class__.suggested:
+            return 'Suggests'
+        if self == self.__class__.recommended:
+            return 'Recommends'
+        if self == self.__class__.required:
+            return 'Requires'
+        raise ValueError
+
+
 def group_by_feature(elffiles):
     features = {}
 
@@ -143,6 +154,52 @@ def generate_rpm(elffiles, stanza, filter):
                 soname = next(iter(note['soname']))  # we take the first — most recommended — soname
                 yield f"{stanza}: {soname}{suffix}"
 
+def rpm_fileattr_generator(args):
+    if args.rpm_features is not None:
+        if not any(fnmatch.fnmatch(args.subpackage, pattern[0])
+                   for pattern in args.rpm_features):
+            # Current subpackage is not listed, nothing to do.
+            # Consume all input as required by the protocol.
+            sys.stdin.read()
+            return
+
+    for file in sys.stdin:
+        file = file.strip()
+        if not file:
+            continue  # ignore empty lines
+
+        elffile = ELFFileReader(file)
+        suffix = '()(64bit)' if elffile.elffile.elfclass == 64 else ''
+
+        first = True
+
+        for note in elffile.notes():
+            # Feature name is optional. Allow this to be matched
+            # by the empty string ('') or a wildcard glob ('*').
+            feature = note.get('feature', '')
+
+            if args.rpm_features is not None:
+                for package_pattern,feature_pattern in args.rpm_features:
+                    if (fnmatch.fnmatch(args.subpackage, package_pattern) and
+                        fnmatch.fnmatch(feature, feature_pattern)):
+                        break
+                else:
+                    # not matched
+                    continue
+            else:
+                # if no mapping, print all features at the suggested level
+                level = Priority[note.get('priority', 'recommended')].rpm_name()
+                if level != args.rpm_fileattr:
+                    continue
+
+            if first:
+                print(f';{file}')
+                first = False
+
+            soname = next(iter(note['soname']))  # we take the first — most recommended — soname
+            print(f'{soname}{suffix}')
+
+
 def make_parser():
     p = argparse.ArgumentParser(
         description=__doc__,
@@ -187,10 +244,28 @@ def make_parser():
         metavar='FEATURE1,FEATURE2',
         help='Generate rpm Recommends for listed features',
     )
+    p.add_argument(
+        '--rpm-fileattr',
+        metavar='TYPE',
+        help='Run as rpm fileattr generator for TYPE dependencies',
+    )
+    p.add_argument(
+        '--subpackage',
+        metavar='NAME',
+        default='',
+        help='Current subpackage NAME',
+    )
+    p.add_argument(
+        '--rpm-features',
+        metavar='SUBPACKAGE:FEATURE,SUBPACKAGE:FEATURE',
+        type=lambda s: [x.split(':', maxsplit=1) for x in s.split(',')],
+        action='extend',
+        help='Specify subpackage:feature mapping',
+    )
     p.add_argument(
         'filenames',
-        nargs='+',
-        metavar='filename',
+        nargs='*',
+        metavar='FILENAME',
         help='Library file to extract notes from',
     )
     p.add_argument(
@@ -207,15 +282,30 @@ def parse_args():
         and not args.sonames
         and args.features is None
         and args.rpm_requires is None
-        and args.rpm_recommends is None):
+        and args.rpm_recommends is None
+        and args.rpm_fileattr is None):
         # Make --raw the default if no action is specified.
         args.raw = True
 
+    if args.rpm_fileattr is not None:
+        if (args.filenames
+            or args.raw
+            or args.features is not None
+            or args.rpm_requires
+            or args.rpm_recommends):
+            raise ValueError('--rpm-generate cannot be combined with most options')
+
+    if args.rpm_fileattr is None and not args.filenames:
+        raise ValueError('At least one positional FILENAME parameter is required')
+
     return args
 
 if __name__ == '__main__':
     args = parse_args()
 
+    if args.rpm_fileattr is not None:
+        sys.exit(rpm_fileattr_generator(args))
+
     elffiles = [ELFFileReader(filename) for filename in args.filenames]
     features = group_by_feature(elffiles)
 
diff --git a/rpm/dlopen_notes.attr b/rpm/dlopen_notes.attr
new file mode 100644
index 0000000000..abe3e8547c
--- /dev/null
+++ b/rpm/dlopen_notes.attr
@@ -0,0 +1,44 @@
+# SPDX-License-Identifier: MIT-0
+#
+# This file is part of the package-notes package.
+#
+#
+# The spec file for a package can specify which features are listed
+# and at which level, using
+# '--rpm-features=SUBPACKAGE1:FEATURE1,SUBPACKAGE2:FEATURE2' option in
+# the '__dlopen_notes_TYPE_opts' macro, where TYPE is one of
+# 'requires', 'recommends', or 'suggests'. The macro should be declared
+# in the spec file using:
+#   %define __dlopen_notes_TYPE_opts SUBPACKAGE:FEATURE…
+# e.g.
+#   %define __dlopen_notes_recommends_opts *:zstd
+#
+# The option accepts multiple comma-separated pairs, and can also be
+# specified multiple times. Both the subpackage name and feature name
+# can be a glob. If configuration is omitted, the priority recommended
+# in the notes is used.
+#
+# The '--subpackage=SUBPACKAGE' option (inserted below) tells the generator
+# which subpackage is being processed.
+#
+# For example, for a package using compression libraries, we can say
+# that the 'package-libs' subpackage shall carry 'Requires' on all the
+# libraries needed for the 'zstd' feature, all subpackages shall carry
+# 'Recommends' on all the libraries needed for the 'gzip' feature, and
+# the 'package' subpackage shall carry 'Suggests' for any feature
+# matching 'lzma' or 'bzip*'.
+#
+#  %define __dlopen_notes_requires_opts   --rpm-features=package-libs:zstd
+#  %define __dlopen_notes_recommends_opts --rpm-features=*:gzip
+#  %define __dlopen_notes_suggests_opts   --rpm-features=package:lzma,package:bzip*
+#
+# To opt out, undefine the %_dlopen_notes_generator macro:
+#  %undefine _dlopen_notes_generator
+
+%_dlopen_notes_generator   %{_bindir}/dlopen-notes
+
+%__dlopen_notes_requires   %{!?_dlopen_notes_generator:true }%{_dlopen_notes_generator} --subpackage='%{name}' --rpm-fileattr=Requires
+%__dlopen_notes_recommends %{!?_dlopen_notes_generator:true }%{_dlopen_notes_generator} --subpackage='%{name}' --rpm-fileattr=Recommends
+%__dlopen_notes_suggests   %{!?_dlopen_notes_generator:true }%{_dlopen_notes_generator} --subpackage='%{name}' --rpm-fileattr=Suggests
+%__dlopen_notes_protocol   multifile
+%__dlopen_notes_magic      ^.*ELF (32|64)-bit.*$
